Updated irr function [issue 98]

yatshunlee opened the pull request:
calculate IRR by calculating all its roots and select the most reasonable IRR among all the real roots

https://numpy.org/doc/stable/reference/generated/numpy.roots.html

yatshunlee commented:
Updated tests/test_financial.py and nump_financial/_financial.py
- test_financial.py: Altered test cases
- _financial.py: Added optional argument `raise_exceptions`

Kai-Striega commented on tests/test_financial.py:
```diff
@@ -764,13 +764,4 @@ def test_irr_no_real_solution_exception(self):
-
-        with pytest.raises(npf.IterationsExceededError):
-            npf.irr(cashflows, maxiter=1, raise_exceptions=True)
+            npf.irr(cashflows, raise_exceptions=True)
```
There's a missing blank line here at the end of the file

Kai-Striega commented on numpy_financial/_financial.py:
```diff
@@ -840,20 +825,40 @@ def irr(values, *, guess=None, tol=1e-12, maxiter=100, raise_exceptions=False):
+    # check sign of all IRR solutions
+    same_sign = np.all(IRR > 0) if IRR[0] > 0 else np.all(IRR < 0)
+    
+    # if the signs of IRR solutions are not the same, first filter potential IRR
+    # by comparing the total positive and negative cash flows.
+    if not same_sign:
+        pos = sum(values[values>0])
+        neg = sum(values[values<0])
+        if pos > neg:
```
I don't understand _why_ you are trying to filter the by the total cashflow. Is this just some heuristic you have picked or is there a more formal process going on here?

yatshunlee replied:
This is just some heuristic I have picked. It works when there are real roots with different signs.

I can give you a simple example and walk you through it:

real roots = -0.5 and 0.1
net sum of the cash flows = +10000

As the net sum > 0, we can kind of exclude the possibility that IRR is -0.5 and keep only 0.1.

Kai-Striega replied:
That makes sense. Thanks for clarifying.

Kai-Striega commented on numpy_financial/_financial.py:
```diff
@@ -840,20 +825,40 @@ def irr(values, *, guess=None, tol=1e-12, maxiter=100, raise_exceptions=False):
+        neg = sum(values[values<0])
+        if pos > neg:
+            IRR = IRR[IRR > 0]
+    else:
+        IRR = IRR[IRR < 0]
```
Should this be indented further?

Kai-Striega commented on numpy_financial/_financial.py:
```diff
@@ -840,20 +825,40 @@ def irr(values, *, guess=None, tol=1e-12, maxiter=100, raise_exceptions=False):
-    g = 1 + guess
+    
+    g = np.roots(values)
+    IRR = np.real(g[np.isreal(g)]) - 1
```
Nitpick: I don't like having all capital variable names, could you rename it to something meaningful?

yatshunlee replied:
IRR -> internal_rate_return

Kai-Striega replied:
That would work. Or maybe something like `irr_` if you wanted to be brief. The rest of the code seems to be brief, although I prefer being explicit in general.

Kai-Striega reviewed (commented):
This looks pretty good. 

Could you add a test case where it would normally have returned a “worse” outcome, e.g. the one in your example?

I'm a bit hesitant to choose a solution for the user: I guess we are doing that implicitly anyway with the iterative solver, so this is an improvement.

yatshunlee commented:
![image](https://github.com/numpy/numpy-financial/assets/69416199/56d92367-353a-425c-8aad-d1458dd468a7)

Interestingly, I just pumped into a situation like this.

The real roots are 0.0 and 0.2136. Considering the sum of cashflow = 0, did Excel give an incorrect solution LOL.

I think we do have to choose a solution for users implicitly. But maybe also include arguments allowing them to choose their options? Like `keep_smallest=True`, `keep_all=False`...

Any suggestion?

Kai-Striega commented:
> did Excel give an incorrect solution?

As you have previously noted, there is more than 1 valid solution. I believe that Excel uses an iterative solver to find a solution, with the initial guess being 10%. My guess is that, as both are mathematically valid, Excel simply chose the first solution it converged too. 

> I think we do have to choose a solution for users implicitly. 

I don't think we _have_ to decide a solution for the user explicitly. We could do it via our existing iterative solver. Although that has weaknesses (as you already know). 

> But maybe also include arguments allowing them to choose their options?

I'm -1 on giving more options to `irr`, it would make the function more complicated (and there are already a lot of options). I've already been considering adding a new function `irr_full_compilment` (or some shorter name) that returns the full complement of solutions and lets the user decide what to do with it.

yatshunlee commented:
It passed the pytest on my local machine:
![image](https://github.com/numpy/numpy-financial/assets/69416199/aed7fd73-5a80-4ca2-ba1f-92cb4f08bedd)

However, I dont know why it doesn't pass the workflow:
![image](https://github.com/numpy/numpy-financial/assets/69416199/63a7ce77-495a-40db-98d5-7cbc49a61c81)

@Kai-Striega may I ask what I can do to resolve this issue?

yatshunlee commented:
> because I don't see why it shows an error here even if the error is actually small here.
> 
> therefore, i removed the test case that I made previously.

I just realized someone had previously made a pull request on changing the _test_financial.py and removed all the assert_almost_equal function, making my original function fail on those newly added sophisicated test cases, with a very small abs difference like 9e-16 which is not 0.

So, I've removed all the test cases that I added previously.

Kai-Striega commented on numpy_financial/_financial.py:
```diff
@@ -727,7 +727,7 @@ def rate(
 
 
-def irr(values, *, guess=None, tol=1e-12, maxiter=100, raise_exceptions=False):
+def irr(values, raise_exceptions=False):
```
The `raise_exceptions` argument should be a [keyword only argument](https://peps.python.org/pep-3102/)

Kai-Striega reviewed (commented):
Sorry for leaving this, I moved states and got distracted :/. I think your solution is better than the iterative approach we use now.

Thanks for taking the time to contribute. I'll play around with it this weekend some more - but it should be approved

Kai-Striega commented:
Changes look good.

I've asked a friend from industry to try and poke some holes into the new method, but I don't expect there to be anything seriously wrong here.

Thanks for taking the initiative!

yatshunlee commented:
> Changes look good.
> 
> I've asked a friend from industry to try and poke some holes into the new method, but I don't expect there to be anything seriously wrong here.
> 
> Thanks for taking the initiative!

Np! My pleasure! I always enjoy making things like this.

Kai-Striega commented:
> Np! My pleasure! I always enjoy making things like this.

That's great to hear. There are plenty of things that will still need to be done on NumPy-financial if you are interested

As an aside, for the future, could you please follow the guide [writing you commit message](https://numpy.org/doc/stable/dev/development_workflow.html#writing-the-commit-message) from the NumPy docs?

yatshunlee commented:
> d to be done on NumPy-financial if you are interested

Yea thanks for letting me know

Kai-Striega commented on numpy_financial/_financial.py:
```diff
@@ -840,20 +825,40 @@ def irr(values, *, guess=None, tol=1e-12, maxiter=100, raise_exceptions=False):
-    d_npv = npv_.deriv()
-    g = 1 + guess
+    
+    g = np.roots(values)
```
From the [roots](https://numpy.org/doc/stable/reference/generated/numpy.roots.html) docs:

> This forms part of the old polynomial API. Since version 1.4, the new polynomial API defined in [numpy.polynomial](https://numpy.org/doc/stable/reference/routines.polynomials.package.html#module-numpy.polynomial) is preferred. A summary of the differences can be found in the [transition guide](https://numpy.org/doc/stable/reference/routines.polynomials.html).

However I'm not sure if it's worth to construct a whole ``Polynomial`` object just for this one calculation...

Kai-Striega replied:
I'm currently re-writing some of the functions to use [numba](https://numba.readthedocs.io/en/stable/index.html), it looks like ``np.roots`` is supported, but the new Polynomial API is not. So I'd prefer this version.

Kai-Striega reviewed (approved):
This looks excellent overall. There's one comment that I will need to think about. I think it's ready to merge. Are there still any changes you would like to do?

Kai-Striega commented:
Merged, Thanks @yatshunlee for taking the time to contribute.
